pkg/queue/manager.go

```diff
@@ -547,3 +594,105 @@ func (m *Manager) reportPendingWorkloads(cqName string, cq ClusterQueue) {
 	}
 	metrics.ReportPendingWorkloads(cqName, active, inadmissible)
 }
+
+func (m *Manager) Start(ctx context.Context) error {
+	defer m.queue.ShutDown()
+
+	queueVisibilityUpdateInterval := time.Duration(m.queueVisibilityUpdateInterval) * time.Second
+
+	for i := 0; i < snapshotWorkers; i++ {
+		go wait.UntilWithContext(ctx, m.takeSnapshot, queueVisibilityUpdateInterval)
+	}
+
+	go wait.UntilWithContext(ctx, m.enqueueSnapshotDelayed, queueVisibilityUpdateInterval)
+
+	<-ctx.Done()
+
+	return nil
+}
+
+func (m *Manager) enqueueSnapshotDelayed(ctx context.Context) {
+	m.RLock()
+	defer m.RUnlock()
+	for cq := range m.clusterQueues {
+		m.queue.Add(cq)
+	}
+}
+
+func (m *Manager) takeSnapshot(ctx context.Context) {
+	for m.processNextSnapshot(ctx) {
+	}
+}
+
+func (m *Manager) processNextSnapshot(ctx context.Context) bool {
+	log := ctrl.LoggerFrom(ctx).WithName("processNextSnapshot")
+
+	key, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+
+	startTime := time.Now()
+	defer func() {
+		log.V(4).Info("Finished snapshot job", "key", key, "elapsed", time.Since(startTime))
+	}()
+
+	cq := m.extractClusterQueue(key)
+	if cq == nil {
+		return false
+	}
+
+	defer m.queue.Done(key)
+
+	workloads := make([]kueue.ClusterQueuePendingWorkload, 0)
+	if elements, ok := cq.Snapshot(m.queueVisibilityClusterQueuesMaxCount); ok {
+		for _, wl := range elements {
+			workloads = append(workloads, kueue.ClusterQueuePendingWorkload{
+				Name:      wl.Name,
+				Namespace: wl.Namespace,
+			})
+		}
+	}
+	m.SetSnapshot(key.(string), workloads)
+
+	return true
+}
+
+func (m *Manager) extractClusterQueue(key interface{}) ClusterQueue {
+	m.RLock()
+	defer m.RUnlock()
+
+	if len(m.clusterQueues) == 0 {
+		return nil
+	}
+
+	if cq := m.clusterQueues[key.(string)]; cq != nil {
+		return cq
+	}
+
+	return nil
+}
+
+func (m *Manager) SetSnapshot(cqName string, workloads []kueue.ClusterQueuePendingWorkload) {
+	m.wm.Lock()
+	defer m.wm.Unlock()
+	m.snapshots[cqName] = workloads
+}
+
+func (m *Manager) GetSnapshot(cqName string) []kueue.ClusterQueuePendingWorkload {
+	m.wm.RLock()
+	defer m.wm.RUnlock()
+	return m.snapshots[cqName]
+}
+
+func (m *Manager) DeleteSnapshot(cq *kueue.ClusterQueue) {
+	m.Lock()
+	defer m.Unlock()
```

Comment:
This defer will only end once `DeleteSnapshot` ends, but we can end it ealier. I suggest to extract a function `getClusterQueue` (if there is no already) and take the lock inside this function.